### ISSUE TITLE
Use catbox-memory for caching for optimization

### ISSUE THREAD

**Issue by rmothilal** [mojaloop/project#602]
## **Goal**:
As a Switch implementer
I want to use catbox-memory to cache the enums on central-ledger
so that the caching mechanism can be separated from the the transport layer ie: Hapi server

**Tasks**:
- [x] Remove current caching implementation on central-ledger hapi server i.e. setup.js 
  - [x] Remove from setup.js
- [x] Add new cache config to default.json and config.js
- [x] Implement code changes to use catbox-memory
  - [x] Create cache.js under src/lib
  - [x] Add current cache enums to the new cache library
- [x] Identify declarations that use _Hard-coded enums accessed through variables_ and access the cache instead, where relevant (especially DB related)
- [ ] ~Add DELETE/REFRESH cache functionality to the API to allow management of cache i.e. look at ALS for example~
- [ ] Cache must be initialised when the server starts up and populated with the following data:
  - [x] All currently cached information
  - [x] Get All Active Participants - participant table
  - [x] Get Active Participant Currency - participantCurrency table using each of the participants from the list of participants returned
  - [ ] ~Store results from src/models/participant/facade.js/getByNameAndCurrency() as its used in a few places or create it from the calls above as it uses those 2 tables~
  - [ ] ~Participant endpoints - participantEndpoint table~
- [x] Step through the prepare and fulfil steps of a transfer to identify and list out places / validations where caching can be used or data can be cached, without affecting integrity [ @oderayi ]
- [x] Refactor enum cache on central-ledger to use catbox-memory instead of the Hapi server 
- [x] Update tests
- [x] Refactor implementation where a particular 'cache' needs to be deleted / cleared / invalidated, for example when a participant gets updated, etc.

**Acceptance Criteria**:
- [x] Standalone catbox-memory is used to cache the enums on central-ledger instead of Hapi plugin
- [x] central-ledger uses catbox-memory instead of the Hapi server for caching
- [x] Cache can be accessed during the session
- [x] Cache must be able to be reset
- [x] All tests pass

**Pull Requests**:
- [x] https://github.com/mojaloop/central-ledger/pull/560 - enums rewritten to direct Catbox memory
- [x] https://github.com/mojaloop/central-ledger/pull/566 - active participants cache
- [x] https://github.com/mojaloop/postman/pull/81 - fixed Postman collection ("hub" vs "Hub" issue)
- [x] https://github.com/mojaloop/central-ledger/pull/576 - refactored participant Cache
- [x] https://github.com/mojaloop/central-ledger/pull/583 - participant-currency caching

**Follow-up**:
- #1214 
- #1215

**Dependencies**:
- #695 

**Accountability**:
- Owner: @ggrg @rmothilal  
- QA/Review: @mdebarros

**Comment by lewisdaly** [comment]
@elnyry @rmothilal I'll take a look at this one if that's ok.

**Comment by lewisdaly** [comment]
It appears that the central-ledger is already using `catbox-memory`. 
From `central-ledger/src/setup.js`:
```js
const server = await new Hapi.Server({
  port,
  cache: [
    {
      provider: {
        constructor: require('catbox-memory'),
        options: {
          partition: 'cache'
        }
      },
      name: 'memCache'
    }
  ],
  . . . 
})
```

Do you mean that we should remove this configuration altogether and implement a custom caching mechanism?

**Comment by ggrg** [comment]
Hi @lewisdaly - yes, this configuration needs to be removed, because the above cache may be used only for http requests and we need pass the request object down to the method that accesses the enums in order to use that enum from the web server cache. Instead, we would like to access cached values from the running node process and to be able to access it not only for http originating requests, but everywhere (e.g. Position Handler, which is triggered by produced kafka message - the http request is against the ml-api-adapter). 

Also, please note that we were also considering adding a shared microservice holding that cache, so we can even access it not only at process level (node), but from all other microservices. @mdebarros, should be able to provide more info in this direction.

**Comment by lewisdaly** [comment]
@elnyry has this become more 'in scope' with the current work on performance improvements?

**Comment by elnyry** [comment]
@lewisdaly yes, in our perf discussions one thing that came up was to have a comprehensive caching strategy.. We want to use this in places where we to do validations on the transfers critical path to decrease the DB reads, etc.. For example, right now we already use this for FSP endpoints as they're needed (multiple times) on every transfers call.. We'd also like to leverage this for FSP details (checking for validity), etc..

I think Rajiv spent some time refining the item to get it ready.. We'd like to do this in parallel as soon as possible, though hopefully we'll get a baseline even before this is merged into master..

**Comment by yosheeck** [comment]
The issue is now blocked by https://github.com/mojaloop/project/issues/1152 as this one is effectively preventing any significant performance improvements until fixed.

**Comment by oderayi** [comment]
Possible caching opportunities on the transfer prepare-fulfill path:

- https://mojaloop.io/documentation/mojaloop-technical-overview/central-ledger/transfers/

Prepare Transfer Request:
--------------------------
- 1.1.2.a Net Debit Cap
    - Position limits can be cached
- 1.1.1.a Prepare Handler Consume 
    - Duplicate check: Unfinalized transfers can be cached and used for duplicate checks first and then DB if not found in the cache.
    - Participants details can be cached and loaded on start up, invalidated and refreshed when changed.
- 1.1.4.a Send Notification to Participant
    - Endpoint details are cached on first fetch only and removed from cache on expiry, then fetched again. 
      This can be improved. Endpoint details can be loaded on start up, and invalidated and refreshed when changed.

Fulfil Transfer Request:
------------------------
- 2.1.1 Fulfil handler consume
    - Duplicate check: Unfinalized transfers can be cached and used for duplicate checks first and then DB if not found in the cache.
    - settlementWindowId fetch: to be discussed
- 1.2.0 Position Handler Consume
    - Position limits can be cached

**Comment by elnyry** [comment]
@oderayi thanks for this.

For this iteration, I think we can do the Limits along with the FSP details. Settlement window ID as you mentioned, needs to discussed as it depends on various factors (such as Settlement model, etc).
